### PR TITLE
Contact system groups by database trigger

### PR DESCRIPTION
**Pull request by rowanseymour**
`Contact.is_test` becomes un-changeable to simplify possible state changes, hence changes to some unit tests to use `Contact.get_test_contact` rather than switching existing contacts into test mode.

### CODE REVIEW

File: temba/contacts/migrations/0018_update_group_triggers.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+from __future__ import unicode_literals
+
+from django.db import models, migrations
+
+# language=SQL
+TRIGGER_SQL = """
+----------------------------------------------------------------------
+-- Trigger procedure to update group count
+----------------------------------------------------------------------
+CREATE OR REPLACE FUNCTION update_group_count() RETURNS TRIGGER AS $$
```
Comment (rowanseymour) [inline temba/contacts/migrations/0018_update_group_triggers.py]: Didn't need to re-include this from the earlier migration but it seems sensible to have as much of our trigger code in as few places as possible. I've even been toying with the idea of pulling it all out into an external SQL file.

File: temba/contacts/migrations/0018_update_group_triggers.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+DROP TRIGGER IF EXISTS when_contact_groups_truncate_then_update_count_trg on contacts_contactgroup_contacts;
+CREATE TRIGGER when_contact_groups_truncate_then_update_count_trg
+  AFTER TRUNCATE ON contacts_contactgroup_contacts
+  EXECUTE PROCEDURE update_group_count();
+
+----------------------------------------------------------------------
+-- Toggle a contact's membership of a system group in their org
+----------------------------------------------------------------------
+CREATE OR REPLACE FUNCTION
+  contact_toggle_system_group(_contact_id INT, _org_id INT, _group_type CHAR(1), _add BOOLEAN)
```
Comment (rowanseymour) [inline temba/contacts/migrations/0018_update_group_triggers.py]: Prefixing params with _ as a convention helps avoid collisions with actual table columns

File: temba/contacts/migrations/0018_update_group_triggers.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+  -- don't do anything if group doesn't exist for some inexplicable reason
+  IF _group_id IS NULL THEN
+    RETURN;
+  END IF;
+
+  IF _add THEN
+    BEGIN
+      INSERT INTO contacts_contactgroup_contacts (contactgroup_id, contact_id) VALUES (_group_id, _contact_id);
+    EXCEPTION WHEN unique_violation THEN
```
Comment (rowanseymour) [inline temba/contacts/migrations/0018_update_group_triggers.py]: Django implements a many-to-many insert as a select followed by an insert. That might be faster if you're frequently trying to insert duplicates (?)... but I'm pretty sure this is faster if we're generally not.

File: temba/contacts/migrations/0018_update_group_triggers.py (hunk 1 above)
Comment (rowanseymour) [inline temba/contacts/migrations/0018_update_group_triggers.py]: Would actually make it easier to read this PR if it were an external sql file and so Github could do some syntax highlighting

File: temba/contacts/migrations/0018_update_group_triggers.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- install for UPDATE on contacts_contact
+DROP TRIGGER IF EXISTS contact_check_update_trg ON contacts_contact;
+CREATE TRIGGER contact_check_update_trg
+   BEFORE UPDATE ON contacts_contact
+   FOR EACH ROW EXECUTE PROCEDURE contact_check_update();
+"""
```
Comment (nicpottier) [inline temba/contacts/migrations/0018_update_group_triggers.py]: Could we optimize the update case to only fire when the appropriate columns are changed?

http://www.postgresql.org/docs/9.0/static/sql-createtrigger.html

**Comment by nicpottier** [comment]
Looks good, though we should consider the tightening the update case to only what we care about.

File: temba/contacts/migrations/0018_update_group_triggers.py (hunk 4 above)
Comment (rowanseymour) [inline temba/contacts/migrations/0018_update_group_triggers.py]: Yeah tho looking at this now am thinking I should update `contact_check_update` to disallow failing, blocking and de-activating of test contacts... in which case there would be 4 columns

Comment (nicpottier) [inline temba/contacts/migrations/0018_update_group_triggers.py]: I'm not all that concerned about the test contact cases, but sure.

**Pull request merged**